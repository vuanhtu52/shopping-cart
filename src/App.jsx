import { Outlet } from "react-router-dom";
import "./App.css";
import NavBar from "./components/Navbar";
import Footer from "./components/Footer";
import ShopContext from "./context/ShopContext";
import { useEffect, useState } from "react";

const App = () => {
    const [cartItems, setCartItems] = useState({});
    const [products, setProducts] = useState([]);

    const addToCart = productId => { 
        setCartItems(prevItems => {
            const updatedItems = {...prevItems};

            if (updatedItems.hasOwnProperty(productId)) {
                updatedItems[productId] += 1;
            } else {
                updatedItems[productId] = 1;
            }

            return updatedItems;
        });
    };

    const addMultipleToCart = (productId, quantity) => {
        setCartItems(prevItems => {
            const updatedItems = {...prevItems};

            if (updatedItems.hasOwnProperty(productId)) {
                updatedItems[productId] += quantity;
            } else {
                updatedItems[productId] = quantity;
            }

            return updatedItems;
        });
    };

    const updateCart = (productId, quantity) => {
        setCartItems(prevItems => {
            const updatedItems = {...prevItems};

            updatedItems[productId] = quantity;

            return updatedItems;
        });
    }

    const removeOneFromCart = productId => {
        console.log("hello");
        setCartItems(prevItems => {
            const updatedItems = {...prevItems};

            if (!updatedItems.hasOwnProperty(productId)) {
                return updatedItems;
            }

            if (updatedItems[productId] > 1) {
                updatedItems[productId] -= 1;
            } else {
                delete updatedItems[productId];
            }

            return updatedItems;
        });
    };

    const removeFromCart = productId => {
        setCartItems(prevItems => {
            const updatedItems = {...prevItems};
            delete updatedItems[productId];
            return updatedItems;
        });
    };

    const getTotalCartItems = () => {
        let total = 0;

        for (let key in cartItems) {
            total += cartItems[key];
        }

        return total;
    };

    const getTotalCartPrice = () => {
        let total = 0;
        for (const product of products) {
            if (cartItems.hasOwnProperty(product.id)) {
                total += product.price * cartItems[product.id];
            }
        }

        total = parseFloat(total.toFixed(2));

        return total;
    };

    useEffect(() => {
        fetch('https://fakestoreapi.com/products')
            .then(res => res.json())
            .then(json => setProducts(json))
    }, []);

    return (
        <ShopContext.Provider value={{ cartItems, products, addToCart, addMultipleToCart, updateCart, removeOneFromCart, removeFromCart, getTotalCartItems, getTotalCartPrice }}>
            <div className="flex flex-col min-h-screen">
                <NavBar />
                <div className="flex-grow">
                    <Outlet />
                </div>
                <Footer />
            </div>
        </ShopContext.Provider>
    );
};

export default App;
