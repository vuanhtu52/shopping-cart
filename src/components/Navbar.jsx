import CartIcon from "../assets/svg/cart.svg";
import MenuIcon from "../assets/svg/menu.svg";
import { useState } from "react";
import ArrowUpIcon from "../assets/svg/arrow-up.svg";
import { Transition } from "@headlessui/react";
import { Link } from "react-router-dom";

const NavBar = () => {
    const [menuOpen, setMenuOpen] = useState(false);

    const toggleMenu = () => {
        setMenuOpen(!menuOpen);
    };

    const closeMenu = () => {
        setMenuOpen(false);
    };

    return (
        <div>
            {/* Top nav bar */}
            <div className="flex min-w-96 items-center justify-between gap-16 bg-zinc-800 p-8 text-white">
                {/* Mobile menu button */}
                <button className="block lg:hidden" onClick={toggleMenu}>
                    <img src={MenuIcon} alt="menu icon" className="w-8" />
                </button>

                {/* Logo */}
                <Link
                    to="/"
                    className="cursor-pointer font-dancing-script text-3xl"
                >
                    Random Shop
                </Link>

                {/* Navigation links */}
                <div className="hidden items-center justify-between gap-16 lg:flex">
                    <Link
                        to="men"
                        className="font-noto-sans underline-offset-8 hover:underline"
                    >
                        Men's Clothing
                    </Link>
                    <Link
                        to="women"
                        className="font-noto-sans underline-offset-8 hover:underline"
                    >
                        Women's Clothing
                    </Link>
                    <Link
                        to="jewelery"
                        className="font-noto-sans underline-offset-8 hover:underline"
                    >
                        Jewelery
                    </Link>
                    <Link
                        to="electronics"
                        className="font-noto-sans underline-offset-8 hover:underline"
                    >
                        Electronics
                    </Link>
                </div>

                {/* Shopping cart button */}
                <Link to="checkout">
                    <img src={CartIcon} alt="cart icon" className="w-8" />
                </Link>
            </div>

            {/* Mobile menu - hidden by default */}
            <Transition
                show={menuOpen}
                enter="transition duration-150 origin-top ease-in-out"
                enterFrom="transform opacity-0 scale-y-0"
                enterTo="transfrom opacity-100 scale-y-100"
                leave="transition duration-150 origin-top ease-in-out"
                leaveFrom="transform opacity-100 scale-y-100"
                leaveTo="transform opacity-0 scale-y-0"
            >
                <div className="fixed flex w-full flex-col bg-zinc-800 p-8 text-white">
                    <Link
                        to="men"
                        className="border-b border-gray-500 py-4 font-noto-sans underline-offset-8 hover:underline"
                    >
                        Men's Clothing
                    </Link>
                    <Link
                        to="women"
                        className="border-b border-gray-500 py-4 font-noto-sans underline-offset-8 hover:underline"
                    >
                        Women's Clothing
                    </Link>
                    <Link
                        to="jewelery"
                        className="border-b border-gray-500 py-4 font-noto-sans underline-offset-8 hover:underline"
                    >
                        Jewelery
                    </Link>
                    <Link
                        to="electronics"
                        className="border-b border-gray-500 py-4 font-noto-sans underline-offset-8 hover:underline"
                    >
                        Electronics
                    </Link>
                    <button
                        className="mt-8 h-8 w-8 self-center p-0"
                        onClick={closeMenu}
                    >
                        <img src={ArrowUpIcon} alt="arrow up icon" />
                    </button>
                </div>
            </Transition>
        </div>
    );
};

export default NavBar;
