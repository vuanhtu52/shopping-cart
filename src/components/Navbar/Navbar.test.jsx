import { render, screen } from "@testing-library/react";
import Navbar from "./Navbar";
import { describe, expect, it } from "vitest";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import App from "../../App";
import NavBar from "./Navbar";

describe("Navbar", () => {
    it("renders navbar", () => {
        const { container } = render(
            <BrowserRouter>
                <Navbar />
            </BrowserRouter>,
        );
        expect(container).toMatchSnapshot();
    });

    // it("hover overs nav link", async () => {
    //     const user = userEvent.setup();

    //     render(
    //         <BrowserRouter>
    //             <App />
    //         </BrowserRouter>,
    //     );

    //     const navLinks = screen.getAllByRole("div");
    //     expect(navLinks[0]).toBeInTheDocument();
    // });
});
