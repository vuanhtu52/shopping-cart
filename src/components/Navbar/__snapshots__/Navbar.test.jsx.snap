// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > renders navbar 1`] = `
<div>
  <div
    class="fixed top-0 w-full"
  >
    <div
      class="flex h-28 min-w-96 items-center justify-between gap-16 bg-zinc-800 p-8 text-white"
    >
      <button
        class="block lg:hidden"
      >
        <img
          alt="menu icon"
          class="w-8"
          src="/src/assets/svg/menu.svg"
        />
      </button>
      <a
        class="cursor-pointer font-dancing-script text-3xl"
        href="/"
      >
        Random Shop
      </a>
      <div
        class="hidden items-center justify-between gap-16 lg:flex"
      >
        <a
          class=" font-noto-sans underline-offset-8 hover:underline"
          href="/men"
        >
          Men's Clothing
        </a>
        <a
          class=" font-noto-sans underline-offset-8 hover:underline"
          href="/women"
        >
          Women's Clothing
        </a>
        <a
          class=" font-noto-sans underline-offset-8 hover:underline"
          href="/jewelery"
        >
          Jewelery
        </a>
        <a
          class=" font-noto-sans underline-offset-8 hover:underline"
          href="/electronics"
        >
          Electronics
        </a>
      </div>
      <a
        href="/checkout"
      >
        <img
          alt="cart icon"
          class="w-8"
          src="/src/assets/svg/cart.svg"
        />
      </a>
    </div>
  </div>
</div>
`;
